SRP based login

est31 opened the pull request:
In order to fix #1943. Comments appreciated. What I have tested so far, and what works:
* Logging in for first time
* Logging in with srp
* Logging in with srp based on the legacy hash
* Changing the password

Adds a bundled `mini-gmp` library to do bignum computations, with default option to use system provided gmp. If it isn't found, or discovery is disabled with `-DUSE_SYSTEM_GMP=0`, bundled lib is used.

nerzhul commented on src/network/clientopcodes.cpp:
```diff
@@ -175,11 +188,26 @@ const ServerCommandFactory serverCommandFactoryTable[TOSERVER_NUM_MSG_TYPES] =
 	{ "TOSERVER_REMOVED_SOUNDS",     1, true }, // 0x3a
 	{ "TOSERVER_NODEMETA_FIELDS",    0, true }, // 0x3b
 	{ "TOSERVER_INVENTORY_FIELDS",   0, true }, // 0x3c
-	{ "TOSERVER_PASSWORD",           0, true }, // 0x3d
```
Don't remove this handler, you need to keep it for compat

est31 replied:
Its `TOSERVER_PASSWORD` not `TOSERVER_PASSWORD_LEGACY`.

nerzhul replied:
Right, then if SRP is implemented this opcode can be changed :)

nerzhul commented on src/network/networkprotocol.h:
```diff
@@ -173,7 +178,12 @@ enum ToClientCommand
-
+	TOCLIENT_ACCESS_DENIED = 0x0A,
+	/*
+		u16 command
```
remove u16 command, we don't care about u16 command with the NetworkPacket design, it's implicit :)

nerzhul replied:
This information was useless with previous design because old design need to add opcode and then datas. With network packet the only useful thing is data

est31 replied:
ok, will do.

nerzhul commented on src/network/networkprotocol.h:
```diff
@@ -589,27 +599,31 @@ enum ToClientCommand
 enum ToServerCommand
 {
-	TOSERVER_INIT = 0x02,
+	TOSERVER_INIT = 0x0f,
```
Why change the opcode ? It's unused at this moment, keep a low opcode please

est31 replied:
I have thought to keep it aligned with existing opcodes, not cause fragmentation. I can change it back if you want.

nerzhul replied:
In fact, if you could add SRP and this new _INIT packet together at 
the lower opcodes it could be good. Legacy upper is not a problem, in 1
year and more we could remove it :)

est31 replied:
I have thought the lower opcodes should be reserved for future new init packets. I don't think we should re-allocate deprecated opcodes, and therefore we need the space.

nerzhul replied:
In fact, no new _INIT packet must be created later, it's the goal with 
this new _INIT packet, when the protocol v25 will be done. The _INIT 
packet shouldn't be changed for years

est31 replied:
Ok, you have convinced me now, I will give the init packet a low id again.

nerzhul commented on src/network/networkprotocol.h:
```diff
@@ -120,11 +120,13 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 		Add TOCLIENT_HELLO for presenting server to client after client
 			presentation
-		Add TOCLIENT_AUTH_ACCEPT to accept connexion from client
+		Add TOCLIENT_AUTH_ACCEPT to accept connecion from client
```
Connection *

nerzhul commented on src/network/serveropcodes.cpp:
```diff
@@ -85,13 +85,28 @@ const ToServerCommandHandler toServerCommandTable[TOSERVER_NUM_MSG_TYPES] =
 	{ "TOSERVER_REMOVED_SOUNDS",           TOSERVER_STATE_INGAME, &Server::handleCommand_RemovedSounds }, // 0x3a
 	{ "TOSERVER_NODEMETA_FIELDS",          TOSERVER_STATE_INGAME, &Server::handleCommand_NodeMetaFields }, // 0x3b
 	{ "TOSERVER_INVENTORY_FIELDS",         TOSERVER_STATE_INGAME, &Server::handleCommand_InventoryFields }, // 0x3c
-	{ "TOSERVER_PASSWORD",                 TOSERVER_STATE_INGAME, &Server::handleCommand_Password }, // 0x3d
```
Don't remove for compat

nerzhul commented on src/network/serveropcodes.cpp:
```diff
@@ -39,7 +39,7 @@ const ToServerCommandHandler toServerCommandTable[TOSERVER_NUM_MSG_TYPES] =
 	null_command_handler, // 0x0d
 	null_command_handler, // 0x0e
-	null_command_handler, // 0x0f
+	{ "TOSERVER_INIT",                     TOSERVER_STATE_NOT_CONNECTED, &Server::handleCommand_Init }, // 0x0f
```
Let it at a low opcode

est31 commented:
I'm not sure, android builds for me. Do you know how to fix it @nerzhul?

nerzhul commented:
I wait this build to look at the issue. Maybe GMP not installed on the slave, waiting the build to tell you

est31 commented:
It should work without GMP being installed. GMP is downloaded manually and cross-compiled. System's GMP can't be used because it isn't compiled for ARM.

nerzhul commented:
How can is this downloaded ?

est31 commented:
[see here](https://github.com/est31/minetest/commit/54f49423257ac6b64cb3c0cee7b1abbeb56e92dd#diff-f38370481954d389961343997a8eb9b5R568), its like all other dependencies.

nerzhul commented:
This should work, because you download it into the makefile... strange, look at this when build finished. Android isn't the priority atm, travis and FreeBSD build should work :)

est31 commented:
I don't have a fix for travis and FreeBSD right now, but I try to get the build systems working.

est31 commented:
Internet [sais](http://stackoverflow.com/questions/7902894/error-when-configuring-gmp) that the m4 package is needed.

nerzhul commented:
m4 is now installed on FreeBSD and the Debian fir Android build

nerzhul commented:
jenkins build please

est31 commented on src/util/string.cpp:
```diff
@@ -195,6 +196,42 @@ std::string translatePassword(std::string playername, std::wstring password)
+	int * len_salt;
+	unsigned char * bytes_v;
+	int * len_v;
+	srp_create_salted_verification_key(SRP_SHA256, SRP_NG_4096, n_playername.c_str(), password.c_str(), password.size, bytes_salt, &len_salt, &bytes_v, len_v);
```
@Zeno- do you have an idea why this doesn't compile?

est31 replied:
OK, found it with the help of smarter clang errors.

Zeno- commented on src/network/clientpackethandler.cpp:
```diff
@@ -1107,3 +1151,7 @@ void Client::handleCommand_EyeOffset(NetworkPacket* pkt)
+void Client::handleCommand_SrpBytesSandB(NetworkPacket* pkt)
+{
+	// TODO SRP
+}
```
:)

est31 replied:
understood.

ShadowNinja commented:
csrp has some unresolved issues and some fixes have been added by forks.  Make sure to add those.

Also, please lowercase player names if you're going to use them in the protocol, it makes implementing case-insensitive player names much cleaner.

est31 commented:
@ShadowNinja I have thought of the case-insensitiveness, [look here](https://github.com/minetest/minetest/commit/6e4cdfab540a09d895704b6f48b522f8f0e6c58f#diff-765e4fc5d8245871d9a5a4db834e4631R204). And about csrp, I'll have a look at those forks.

est31 commented:
Ok, I've added some of the commits from the forks. Added missing `mod N` for B and S calculation, also H_nn padding (with buffer underflow protection). We abort nicely when A == 0 (mod N). The gcc fix has already been done by me independently, as I've stumbled across that, too. We should be compatible with RFC 5054, but I haven't tested the test vectors yet. See est31/csrp-gmp@4ce157985c972411f7da9ad8afbceafdaf54532e for the changes, which also have been applied to the file in the PR.

Zeno- commented:
I really wish this was possible without adding a MP lib to the source tree :(

est31 commented:
It is possible, and using system provided gmp is even recommended, as mini-gmp is horribly slow in comparison. However, if mini-gmp gets removed, the system gmp dependency will become hard instead of being soft. That's the only reason its in there.

nerzhul commented:
If it's possible to have an option in cmake to choose embedded or system lib it could ne great

est31 commented:
There is one, I only have to document it.

nerzhul commented:
Perfect

Zeno- commented on src/util/string.cpp:
```diff
@@ -196,6 +197,93 @@ std::string translatePassword(const std::string &playername,
+	char * salt = NULL;
+	std::string ret_val = getSRPVerifier(playername,
+		password, (const char **) &salt, 0);
+	if (!salt)
```
You don't need to check if salt is not NULL  (i.e. free(NULL); does not cause a problem)

est31 replied:
Oh, thats stupid.

est31 commented:
Ready for review @kwolekr.

est31 commented on src/client.cpp:
```diff
@@ -943,6 +951,41 @@ void Client::interact(u8 action, const PointedThing& pointed)
+}
+
+
+AuthMechanism Client::get_auth_mech(const u32 mechs)
```
TODO: name this chose_auth_mech

kwolekr replied:
Why do you need to add the const qualifier to an integer parameter clearly passed by value??

kwolekr replied:
Not really sure what the purpose of this method is to begin with - I assume there can be multiple mechanisms at once, and this chooses which to prefer?

kwolekr replied:
Also, this shouldn't this method name be camelCase?  lower_unix_underscore_style() is for functions not part of any class.

est31 replied:
Removing const (yes it doesn't make sense you are right), and making it camelCase `choseAuthMech`. Yes, server can send multiple allowed auth mechanisms. This future proof-ness doesn't come at any cost.

kwolekr commented on src/client.cpp:
```diff
@@ -943,6 +951,41 @@ void Client::interact(u8 action, const PointedThing& pointed)
+		case AUTH_MECHANISM_SRP:
+		case AUTH_MECHANISM_LEGACY_PASSWORD:
+			srp_user_delete((SRPUser *) m_auth_data);
+			m_auth_data = 0;
```
If you want to clear a pointer, set it to NULL, not 0!
0 tells me "this is a number"
NULL tells me "this was a pointer to something"

kwolekr commented on src/client.cpp:
```diff
@@ -943,6 +951,41 @@ void Client::interact(u8 action, const PointedThing& pointed)
+
+AuthMechanism Client::get_auth_mech(const u32 mechs)
+{
+	std::set<AuthMechanism> auth_methods;
```
Isn't this unused?

est31 replied:
ah thanks forgotten

Zeno- commented on src/client.cpp:
```diff
@@ -956,6 +999,73 @@ void Client::sendLegacyInit(const char* playerName, const char* playerPassword)
+
+	switch (chosen_auth_mechanism) {
+		case AUTH_MECHANISM_FIRST_SRP: {
+			m_authstate = 1;
```
Personally I'd have each of these cases as a function

kwolekr replied:
What is this magic constant '1', and what does it mean?

Zeno- commented on src/client.cpp:
```diff
@@ -956,6 +999,73 @@ void Client::sendLegacyInit(const char* playerName, const char* playerPassword)
+			resp_pkt.putCString(bytes_v, len_v);
+			free((char *) salt);
+			free((char *) bytes_v);
+			resp_pkt << ((m_password == "") ? (u8) 1 : (u8) 0);
```
Casting the whole expression might be neater. e.g. `resp_pkt << (u8)(m_password == "" ? 1 : 0);

est31 replied:
yea, agree.

kwolekr commented on src/client.cpp:
```diff
@@ -22,10 +22,12 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include <sstream>
 #include <IFileSystem.h>
 #include "jthread/jmutexautolock.h"
+#include "util/auth.h"
```
Should auth.cpp/h and srp.c/h really be in util?? They do take on larger roles than utility functions, you know.

est31 replied:
so you mean in src itself? Its so cramped there... network isn't good either because its not low level network related (like networkpacket) nor spanning a wide range of topics (like clientpackethandler).

kwolekr commented on src/client.cpp:
```diff
@@ -943,6 +951,41 @@ void Client::interact(u8 action, const PointedThing& pointed)
+			break;
+		case AUTH_MECHANISM_SRP:
+		case AUTH_MECHANISM_LEGACY_PASSWORD:
+			srp_user_delete((SRPUser *) m_auth_data);
```
The rest of the code has no spaces between the cast and the expression being casted.

Zeno- commented on src/clientiface.h:
```diff
@@ -201,10 +230,24 @@ class RemoteClient
+	u32 allowed_auth_mechs;
+	u32 allowed_sudo_mechs;
+
+	bool is_sudo_mech_allowed(int mech)
```
Pedantically, this would preferably be `bool is_sudo_mech_allowed(unsigned mech)`

kwolekr commented on src/client.cpp:
```diff
@@ -956,6 +999,73 @@ void Client::sendLegacyInit(const char* playerName, const char* playerPassword)
+				&len_salt, &bytes_v, &len_v);
+			resp_pkt.putCString(salt, len_salt);
+			resp_pkt.putCString(bytes_v, len_v);
+			free((char *) salt);
```
You don't need to cast pointers to anything with free()

est31 replied:
gives me
```error: invalid conversion from ‘const void*’ to ‘void*’```

kwolekr commented on src/client.cpp:
```diff
@@ -956,6 +999,73 @@ void Client::sendLegacyInit(const char* playerName, const char* playerPassword)
+			int len_salt, len_v;
+			salt = NULL;
+			getSRPVerifier(getName(), m_password, &salt,
+				&len_salt, &bytes_v, &len_v);
```
Lengths or sizes of things should be size_t

kwolekr commented on src/client.cpp:
```diff
@@ -956,6 +999,73 @@ void Client::sendLegacyInit(const char* playerName, const char* playerPassword)
+			resp_pkt << ((m_password == "") ? (u8) 1 : (u8) 0);
+			Send(&resp_pkt);
+			break;
+		} case AUTH_MECHANISM_SRP:
```
Why do you have the next case on the same line as the last bracket??

kwolekr commented on src/client.cpp:
```diff
@@ -956,6 +999,73 @@ void Client::sendLegacyInit(const char* playerName, const char* playerPassword)
+		} case AUTH_MECHANISM_SRP:
+		case AUTH_MECHANISM_LEGACY_PASSWORD: {
+			m_authstate = 1;
+			u8 based_on = 1;
```
what does based_on do?

est31 replied:
Consult [networkprotocol.h](https://github.com/minetest/minetest/pull/2620/files#diff-d6c71fb6c2c5e2ad979f39a0422f7786R893). It includes documentation of all packets ever used by minetest.

kwolekr commented on src/client.cpp:
```diff
@@ -956,6 +999,73 @@ void Client::sendLegacyInit(const char* playerName, const char* playerPassword)
+			}
+
+			std::string playername_u = getName();
+			std::transform(playername_u.begin(), playername_u.end(),
```
Cool use of std::transform, but wouldn't it be better to add an uppercase() function in util/string.h like there is lowercase()?

est31 replied:
ah didn't know of that function, will use `lowercase` instead.

kwolekr commented on src/client.h:
```diff
@@ -542,11 +545,21 @@ class Client : public con::PeerHandler, public InventoryManager, public IGameDef
 
+	// Helper function
+	inline std::string getName()
+	{ return m_env.getLocalPlayer()->getName(); }
```
Shouldn't this be something more like getLocalPlayerName to be more descriptive??

est31 replied:
I think both are ok, but if I'll change it.

kwolekr commented on src/client.h:
```diff
@@ -585,9 +600,24 @@ class Client : public con::PeerHandler, public InventoryManager, public IGameDef
+	// If set, this will be sent (and cleared) upon a TOCLIENT_ACCEPT_SUDO_MODE
+	std::string m_new_password;
+	// Usable by auth mechanisms.
+	int m_authstate;
```
What does m_authstate represent??

kwolekr commented on src/client.h:
```diff
@@ -585,9 +600,24 @@ class Client : public con::PeerHandler, public InventoryManager, public IGameDef
+	std::string m_new_password;
+	// Usable by auth mechanisms.
+	int m_authstate;
+	int m_chosen_auth_mech;
```
Shouldn't this be AuthMechanism?

kwolekr commented on src/client.h:
```diff
@@ -585,9 +600,24 @@ class Client : public con::PeerHandler, public InventoryManager, public IGameDef
+	// Usable by auth mechanisms.
+	int m_authstate;
+	int m_chosen_auth_mech;
+	void * m_auth_data;
```
I would've personally implemented this a bit differently.  Here we have a decent case for making AuthMechanism its own virtual base class (read: interface) with each auth mechanism an implementation of this.  It would store the SRPData or whatever inside of this object.

est31 replied:
We could move some things there, but not all. auth_data of course, and m_authstate, but packet handlers better stay in their files. I don't know of any methods that would fit there, except for initialisation and deinitialisation.

kwolekr commented on src/clientiface.cpp:
```diff
@@ -427,10 +428,12 @@ void RemoteClient::notifyEvent(ClientStateEvent event)
-		{
-		case CSE_Init:
-			m_state = CS_InitSent;
+		switch(event) {
```
There should be a space between switch and (event)

kwolekr commented on src/clientiface.cpp:
```diff
@@ -447,7 +450,32 @@ void RemoteClient::notifyEvent(ClientStateEvent event)
+		{
+		case CSE_AuthAccept:
+			m_state = CS_AwaitingInit2;
+			if ((chosen_mech == AUTH_MECHANISM_SRP)
```
I question if details such as authentication choosing and deleting the verifier should be here in this state transition function, or elsewhere

kwolekr commented on src/clientiface.h:
```diff
@@ -201,10 +230,24 @@ class RemoteClient
 		peer_id(PEER_ID_INEXISTENT),
 		serialization_version(SER_FMT_VER_INVALID),
 		net_proto_version(0),
+		chosen_mech(0), // AUTH_MECHANISM_NONE,
```
If this is initialized to AUTH_MECHANISM_NONE, then what's wrong with just setting it to AUTH_MECHANISM_NONE instead of 0?

nerzhul replied:
I prefer it too

kwolekr commented on src/clientiface.h:
```diff
@@ -201,10 +230,24 @@ class RemoteClient
 		serialization_version(SER_FMT_VER_INVALID),
 		net_proto_version(0),
+		chosen_mech(0), // AUTH_MECHANISM_NONE,
+		auth_data(0),
```
again, here, null pointers should be NULL

kwolekr commented on src/network/clientpackethandler.cpp:
```diff
@@ -44,10 +45,16 @@ void Client::handleCommand_Hello(NetworkPacket* pkt)
-	*pkt >> deployed;
+	u32 auth_mechs;
+	std::string username_legacy; // for case insensitivity
+	*pkt >> deployed >> (u32&) auth_mechs >> username_legacy;
```
Why cast it to a u32 reference??

est31 replied:
you can't do `(u32)` for `>>`, therefore I did `(u32&)`. Of course, this doesn't make sense anymore (has been added for debugging), so I'll remove it.

kwolekr commented on src/network/clientpackethandler.cpp:
```diff
@@ -44,10 +45,16 @@ void Client::handleCommand_Hello(NetworkPacket* pkt)
 	infostream << "Client: TOCLIENT_HELLO received with "
-			"deployed=" << ((int)deployed & 0xff) << std::endl;
+			"deployed=" << ((int)deployed & 0xff) << ", auth_mechs="
+			<< auth_mechs <<", chosen=" << chosen_auth_mechanism << std::endl;
```
You're missing a space between << and ", chosen".

kwolekr commented on src/network/clientpackethandler.cpp:
```diff
@@ -56,14 +63,45 @@ void Client::handleCommand_Hello(NetworkPacket* pkt)
+				|| (m_chosen_auth_mech == AUTH_MECHANISM_LEGACY_PASSWORD)) {
+			srp_user_delete((SRPUser *) m_auth_data);
+			m_auth_data = 0;
+			}
```
This closing bracket is at the wrong indentation level!

kwolekr commented on src/network/networkpacket.cpp:
```diff
@@ -81,6 +81,24 @@ void NetworkPacket::putRawString(const char* src, u32 len)
 	m_read_offset += len;
 }
 
+void NetworkPacket::putCString(const char* src, u32 len)
```
Why is this needed at all???

est31 replied:
I've copied from `NetworkPacket::operator<<(std::string src)`.

kwolekr commented on src/network/networkpacket.cpp:
```diff
@@ -81,6 +81,24 @@ void NetworkPacket::putRawString(const char* src, u32 len)
 
+void NetworkPacket::putCString(const char* src, u32 len)
+{
+	u16 msgsize = len;
```
What's the point behind this assignment here?

est31 replied:
I've copied from `NetworkPacket::operator<<(std::string src)`.

kwolekr commented on src/network/networkpacket.cpp:
```diff
@@ -81,6 +81,24 @@ void NetworkPacket::putRawString(const char* src, u32 len)
+{
+	u16 msgsize = len;
+	if (msgsize > 0xFFFF) {
+		msgsize = 0xFFFF;
```
Silently truncating data is bad...

sfan5 replied:
Won't the compiler just optimize the *if* away because an u16 can't be `> 0xffff`?

est31 replied:
even if it won't, this if is non-functional, isn't it?

kwolekr commented on src/network/serverpackethandler.cpp:
```diff
@@ -1223,33 +1192,33 @@ void Server::handleCommand_Breath(NetworkPacket* pkt)
+	RemoteClient* client = getClient(pkt->getPeerId(), CS_Created);
+	if (client->net_proto_version >= 25) {
+		infostream << "Server::handleCommand_Password(): Denying change: "
+		" Client protocol version for peer_id=" << pkt->getPeerId()
```
The line continuations need to be indented too!

est31 commented:
this inherits high prio status from #1943.

est31 commented:
As already noted in irc, this is ready for review.

Zeno- commented:
+1 from me.

nerzhul commented:
+1 for me too.

est31 commented:
181f7baa453c58d4070de7196fd74663110946a8
82e35edff52d88dcd64a9bfc9d2c4c93f1341b78
